src/core/data/ions/ions.h

```diff
@@ -234,62 +234,253 @@ namespace core
-            auto ix1 = layout.physicalEndIndex(QtyCentering::primal, Direction::X);
-            auto ix2 = layout.ghostEndIndex(QtyCentering::primal, Direction::X);
+        }
 
-            auto set = [](auto& pop, auto start, auto stop) {
-                for (auto i = start; i < stop; ++i)
+        template<typename Field, typename Value>
+        void set(Field& field, Value const& value)
+        {
+            for (std::size_t i = coords[0]; i < coords[0] + size; i++)
+            {
+                if constexpr (dim >= 2)
                 {
-                    pop.density()(i) = NAN;
-                    for (auto& [id, type] : Components::componentMap)
-                        pop.flux().getComponent(type)(i) = NAN;
+                    for (std::size_t j = coords[1]; j < coords[1] + size; j++)
+                    {
+                        if constexpr (dim >= 3)
+                        {
+                            for (std::size_t k = coords[2]; k < coords[2] + size; k++)
+                                field(i, j, k) = value;
+                        }
+                        else
+                        {
+                            field(i, j) = value;
+                        }
+                    }
                 }
-            };
+                else
+                {
+                    field(i) = value;
+                }
+            }
+        }
 
+    private:
+        std::size_t size;
+        std::array<std::size_t, dim> coords;
+    };
 
-            auto copy = [ix0, ix1](auto& pop) {
-                pop.density()(ix0 - 1) = pop.density()(ix0);
-                pop.density()(ix1 + 1) = pop.density()(ix1);
 
+    template<typename Ions, typename GridLayout>
+    void fixMomentGhosts_nd(DimConst<2>, Ions& ions, GridLayout const& layout)
+    {
+        auto set = [](auto& pop, auto x, auto start, auto stop) {
+            for (auto i = start; i < stop; ++i)
+            {
+                pop.density()(x, i) = NAN;
                 for (auto& [id, type] : Components::componentMap)
-                {
-                    pop.flux().getComponent(type)(ix0 - 1) = pop.flux().getComponent(type)(ix0);
-                    pop.flux().getComponent(type)(ix1 + 1) = pop.flux().getComponent(type)(ix1);
-                }
-            };
+                    pop.flux().getComponent(type)(x, i) = NAN;
+            }
+        };
+
+        auto ix0 = layout.physicalStartIndex(QtyCentering::primal, Direction::X);
+        auto ix1 = layout.physicalEndIndex(QtyCentering::primal, Direction::X);
+        auto ix2 = layout.ghostEndIndex(QtyCentering::primal, Direction::X);
+
+        auto iy0 = layout.physicalStartIndex(QtyCentering::primal, Direction::Y);
+        auto iy1 = layout.physicalEndIndex(QtyCentering::primal, Direction::Y);
+        auto iy2 = layout.ghostEndIndex(QtyCentering::primal, Direction::Y);
+
+        auto copy = [&](auto& field) {
+            // topLeft
```

Comment:
convention is that 0 left/bottom and 1 is right/top

Reply:
ok I wasn't sure